src/main/java/net/snowflake/client/core/SFJsonResultSet.java

```diff
@@ -250,4 +256,242 @@ public Date getDate(int columnIndex, TimeZone tz) throws SFException {
+    } catch (JsonProcessingException e) {
+      throw new SFException(e, ErrorCode.INVALID_STRUCT_DATA);
+    }
+  }
+
+  private static JsonStringToTypeConverter createArrayConverter() {
+    JsonStringToTypeConverter arrayConverter =
+        value -> {
+          try {
+            return OBJECT_MAPPER.readValue(value, HashMap[].class);
+          } catch (JsonProcessingException e) {
+            throw new SFException(e, ErrorCode.INVALID_STRUCT_DATA);
+          }
+        };
+    return arrayConverter;
+  }
+
+  private static JsonStringToTypeConverter createStructConverter() {
+    JsonStringToTypeConverter structConverter =
+        value -> {
+          try {
+            return OBJECT_MAPPER.readValue(value, Map.class);
+          } catch (JsonProcessingException e) {
+            throw new SFException(e, ErrorCode.INVALID_STRUCT_DATA);
+          }
+        };
+    return structConverter;
+  }
+
+  private JsonStringToTypeConverter createTimestampConverter(
+      int columnSubType, int columnType, int scale) {
+    JsonStringToTypeConverter timestampConverter =
+        value -> {
+          Timestamp result = getTimestampFromType(columnSubType, (String) value, session);
+          if (result != null) {
+            return result;
+          }
+          return converters
+              .getDateTimeConverter()
+              .getTimestamp(value, columnType, columnSubType, null, scale);
+        };
+    return timestampConverter;
+  }
+
+  private JsonStringToTypeConverter createTimeConverter() {
+    JsonStringToTypeConverter timeConverter =
+        value -> {
+          SnowflakeDateTimeFormat formatter =
+              SnowflakeDateTimeFormat.fromSqlFormat(
+                  (String) session.getCommonParameters().get("TIME_OUTPUT_FORMAT"));
+          SFTimestamp timestamp = formatter.parse((String) value);
+          return Time.valueOf(
+              Instant.ofEpochMilli(timestamp.getTime()).atZone(ZoneOffset.UTC).toLocalTime());
+        };
+    return timeConverter;
+  }
+
+  private JsonStringToTypeConverter createDateConverter() {
+    JsonStringToTypeConverter dateConverter =
+        value -> {
+          SnowflakeDateTimeFormat formatter =
+              SnowflakeDateTimeFormat.fromSqlFormat(
+                  (String) session.getCommonParameters().get("DATE_OUTPUT_FORMAT"));
+          SFTimestamp timestamp = formatter.parse((String) value);
+          return Date.valueOf(
+              Instant.ofEpochMilli(timestamp.getTime()).atZone(ZoneOffset.UTC).toLocalDate());
+        };
+    return dateConverter;
+  }
+
+  private Object[] convertToNumericArray(
+      Iterator nodeElements, JsonStringToTypeConverter bigIntConverter) {
+    AtomicInteger bigDecimalCount = new AtomicInteger();
+    Object[] elements =
+        getStream(nodeElements, bigIntConverter)
+            .map(
+                elem -> {
+                  if (elem instanceof BigDecimal) {
+                    bigDecimalCount.incrementAndGet();
```

Comment:
What do you think about such approach:
```
list = getStream(...).toList()
useBigDecimals = getStream(...).filter(elem -> elem instanceof BigDecimal).findAny().isPresent()
map to proper array
```
We don't need to go over whole collection.

Reply:
but nodeElements is a stream too. We can't create stream on this few times.